lua/plugins.lua

```diff
@@ -59,10 +59,7 @@ return require("packer").startup(function(use)
   -- Autopairs
   use {
     "windwp/nvim-autopairs",
-    -- event = "InsertEnter",
-    config = function()
-      require "core.autopairs"
-    end,
+    config = require("core.autopairs").setup(),
```

Comment:
AFAIK, this needs to be a either a string or keep the `function() .. end` wrapper.

Reply:
Nah, it should be `require("core.autopairs").setup` here it's called directly.
Also, config = setup triggers me quick a bit (not related to your work)

Reply:
I tested.  This style works, but I'll wrap it in a function just to be safe.  You're right, I'm not sure if this will cause problems reloading a config.  I'll also have to change what I return in autopairs.lua.

Reply:
I don't understand how it can work the way you expect it to, you assign to config the return value of the function, yes it is called, but not from packer once the plugin is loaded, since config is probably = nil.

Reply:
I tried  ``` require("core.autopairs").setup ``` without calling it directly but it didn't work for the user config.

Reply:
You should try when it's lazyloaded. I thought about it that `Packer` needs to copy that line verbatim in that case, and it's not actually calling inline. I'll do some testing with a few plugins that I have and see if there's any truth to it 😄

Reply:
Okay so according to the docs: 
>     config = string or function, -- Specifies code to run after this plugin is loaded.
So my guess is that it didn't quite work for us when it was a normal `require('some-config')` since that does not necessarily return a function callback. 

Use whichever you think fits, I don't have a preference either way as long as it works!

Reply:
So these ways work, but require you to recompile any time you change from the user config to the default config.

```
config = require("core.autopairs").setup
config = function() require("core.autopairs").setup end
```

This way works, and does not require the user to recompile

```
config = require("core.autopairs").setup(),
```

Reply:
I agree with what you said above, but the problem is that by using the second option, the function is called when plugins.lua is interpreted.
What you want is to only call it when the plugin is loaded (packer takes care of this by calling the config callback).

So for instance, If I don't have it installed yet (first time user for instance), First the call to setup will fail because of `pcall(require, "nvim-autopairs")`, then after running PackerInstall, because config = nil `  if not status_ok then return end` it will stay unconfigured until the editor is restarted.

Reply:
Aha!  That's why those guard clauses are there; for the case where it hasn't been installed yet.  :)  Thanks for clarifying that.  Ok.  I'll pass a single callback to packer.

Reply:
It's more verbose now but it works.  And it fulfills the packer contract.   I tried to have the logic happen in the autopairs file but if I put it there, PackerCompile would fail to link properly to the correct callback function.  

```
  -- Autopairs
  use {
    "windwp/nvim-autopairs",
    config = function()
      if O.plugin.autopairs.config then
        O.plugin.autopairs.config()
      else
        require("core.autopairs").setup()
      end
    end,
  }

```

Reply:
Alright, I think I got it now.  Thanks for bearing with me